### PR TITLE
kubeadm: add support for Json6902 Patches

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This PR fixes limitations of the first implementation of kustomize support in kubeadm (https://github.com/kubernetes/kubernetes/pull/80905)
- support for customization stored in a git repository
- support for patchJson6902 patches

**Special notes for your reviewer**:
Now kubeadm support for kustomize can work with both when the user provides a kustomization file or not.
Support for working with a kustomization file wad not included in the initial design described in the KEP, but the implementation was simpler than expected and it enables support for patchJson6902 and many other kustomize features, so IMO this should go in the alpha release.

**Does this PR introduce a user-facing change?**:
```release-note
NONE
``` 

/sig cluster-lifecycle
/priority important-soon
/assign @rosti @yagonobre @neolit123 @ereslibre

### CODE REVIEW

File: cmd/kubeadm/app/util/kustomize/json6902.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+type json6902 struct {
+	// Target refers to a Kubernetes object that the json patch will be applied to
+	*patch.Target
+
+	// Patch contain the json patch as a string
+	Patch string
+}
+
+// stategicMergeSlice is a slice of json6902 patches.
+type json6902Slice []*json6902
```
Comment (neolit123) [inline cmd/kubeadm/app/util/kustomize/json6902.go]: comment above needs adjustment?

File: cmd/kubeadm/app/util/kustomize/BUILD [hunk 2]
```diff
@@ -3,21 +3,25 @@ load("@io_bazel_rules_go//go:def.bzl", "go_library", "go_test")
 go_library(
     name = "go_default_library",
     srcs = [
+        "json6902.go",
         "kustomize.go",
-        "unstructured.go",
+        "stategicMerge.go",
```
Comment (neolit123) [inline cmd/kubeadm/app/util/kustomize/BUILD]: there is a typo in the filename (for the _test.go) file too.

File: cmd/kubeadm/app/util/kustomize/stategicMerge.go [hunk 3]
```diff
@@ -26,41 +26,33 @@ import (
...
-		if err != nil {
-			return nil, errors.Wrapf(err, "convert %q to Unstructured failed", path)
-		}
+// stategicMergeSlice is a slice of strategic merge patches.
+// Unstructured objects are used to represent strategic merge patches of any group/version/kind.
+type stategicMergeSlice []*unstructured.Unstructured
 
-		result = append(result, res...)
+// newStategicMergeSliceFromFile returns a slice of strategic merge patches from a file
+func newStategicMergeSliceFromFile(loader ifc.Loader, path string) (stategicMergeSlice, error) {
```
Comment (neolit123) [inline cmd/kubeadm/app/util/kustomize/stategicMerge.go]: Stategic -> Strategic

File: cmd/kubeadm/app/util/kustomize/stategicMerge.go [hunk 4]
```diff
@@ -26,41 +26,33 @@ import (
...
+	content, err := loader.Load(path)
+	if err != nil {
+		return nil, errors.Wrapf(err, "load from path %q failed", path)
 	}
-	return result, nil
+	res, err := newStategicMergeSliceFromBytes(content)
```
Comment (neolit123) [inline cmd/kubeadm/app/util/kustomize/stategicMerge.go]: Stategic -> Strategic

Comment (neolit123) [reply cmd/kubeadm/app/util/kustomize/stategicMerge.go]: looks like the typo is present in a lot of places.

**Comment by neolit123** [comment]
@fabriziopandini i did a first pass and found some typos mostly.
i must admit i was expecting less boilerplate for the kustomize / kubeadm integration, but SGTM as long as you are happy with this.

**Comment by fabriziopandini** [comment]
@neolit123 thanks for the feedback!
Everything should be addressed.

> I was expecting less boilerplate for the kustomize / kubeadm integration

If we want kustomize to be applied consistently with kubeadm phases some scaffolding is necessary. Additionally, ATM we are also constrained to not expand required dependencies; probably when kubeadm moves in its own repository we can refactor this a little bit

**Comment by neolit123** [comment]
/lgtm
/hold
(waiting for more reviews, feel free to unhold on lazy consensus)

File: cmd/kubeadm/app/util/kustomize/kustomize.go [hunk 5]
```diff
@@ -145,33 +201,45 @@ func (km *Manager) Kustomize(res []byte) ([]byte, error) {
...
+	for i, p := range strategicMerge {
+		b, err := yaml2.Marshal(p)
 		if err != nil {
 			return nil, err
 		}
-
-		name := fmt.Sprintf("resource-%d.json", i)
+		name := fmt.Sprintf("patch-%d.yaml", i)
 		_ = memFS.WriteFile(filepath.Join(fakeDir, name), b)
 		fmt.Fprintf(&kustomization, " - %s\n", name)
```
Comment (rosti) [inline cmd/kubeadm/app/util/kustomize/kustomize.go]: This line needs to be removed along with `var kustomization bytes.Buffer` above.

**Review by rosti — COMMENTED** [review]
Thanks @fabriziopandini !
Some tiny nits, but the idea turns out quite nice!

**Comment by fabriziopandini** [comment]
@rosti thanks for the feedback!
Commit addressed

**Comment by ereslibre** [comment]
Thank you @fabriziopandini!

/lgtm
/hold

Feel free to remove the hold.

**Review by rosti — APPROVED** [review]
Thanks @fabriziopandini !
/lgtm
/hold cancel

**Pull request merged**